functorch/functorch/_src/partitioners.py

```diff
@@ -164,6 +165,76 @@ def default_partition(
 
     return _extract_fwd_bwd_modules(joint_module, saved_values)
 
+def move_input_mutations_into_submodule(fx_g: fx.GraphModule):
+    mutable_nodes: List[Node] = []  # the copy_() and as_strided_() nodes, that we'll delete move to a submodule.
+    node_map: Dict[Node, Node] = {}
+    node_to_placeholder: Dict[Node, Node] = OrderedDict()  # mapping of nodes from old graph to placeholder in new graph
+    mutation_subgraph = fx.Graph()
+
+    def remap_inputs(x):
+        if x.op == "get_attr":
+            pass
+        if x in node_map:
+            return node_map[x]
+        if x not in node_to_placeholder:
+            # Add the copy_() arg inputs as a submodule input.
+            placeholder_node = mutation_subgraph.placeholder(x.name, type_expr=x.type)
+            placeholder_node.meta = copy.copy(x.meta)
+            node_to_placeholder[x] = placeholder_node
+        return node_to_placeholder[x]
+
+    inpt_nodes = [n for n in fx_g.graph.nodes if n.op == 'placeholder']
+    for n in fx_g.graph.nodes:
+        if not isinstance(n.target, torch._ops.OpOverload):
+            continue
+        mutable_schema_args = [
+            a for a in n.target._schema.arguments
+            if a.alias_info is not None and a.alias_info.is_write
+        ]
+        if len(mutable_schema_args) == 0:
+            continue
+        # First some basic assertions.
+        # This graph should be entirely functional, but might contain
+        # some copy_() mutation nodes, that copy data back into program inputs.
+        # Any mutable ops that we find should correspond to copy_() nodes.
+        assert n.target == torch.ops.aten.copy_.default or \
+               n.target == torch.ops.aten.as_strided_.default, \
+               "Given a non-functionalized graph."
+        # Add the copy_() node to the submodule, remove it from the original graph.
+        mutable_nodes.append(n)
+        new_node = mutation_subgraph.node_copy(n, remap_inputs)
+        node_map[n] = new_node
+    # Add the submodule to the end of the original graph
+    mutation_subgraph.lint()
+    mutation_submodule = lift_subgraph_as_module(fx_g, mutation_subgraph)
+    submodule_name = 'input_mutation_submodule'
+    fx_g.add_submodule(submodule_name, mutation_submodule)
+
+    # Create a call_module node in main graph.
+    module_node = fx_g.graph.call_module(
+        submodule_name,
+        args=tuple(node_to_placeholder.keys()),
+        kwargs=None)
```

Comment:
I know we debated a bit possible representations, but this particular rep is actually not one that I had been thinking about. Mutation be in a submodule *in the original* graph is not great, because it still means the outer graph is *not functional*! (If you call a mutating function inside a graph, that makes your graph mutating.) I feel like there's an obligation for this submodule call to (somehow) not live in the Graph itself, as that makes it vulnerable to DCE again.

Reply:
Hmmmm I see. The alternative is probably just to leave it out of the `fx.Graph` (but still keep it in the `GraphModule`), and then manually cal the submodule later on in AOTAutograd, after the compiled forward gets executed? That doesn't seem too bad.

I guess my thought was that anything that operates on the `fx.Graph` (DCE + compilers) would just see a custom submodule and know to treat it as an opaque object. But that's probably not totally right - we can't really enforce that a graph pass won't move the submodule around earlier in the graph (which would be wrong).

Reply:
although, if I want to be able to invoke the submodule separately outside of the graph, I'll need to somehow get the inputs for the submodule - probably by updating the original graph to make them additional outputs. Extra complexity (since I'll also need to remove the pytree stuff to do that), but doesn't seem too bad.

Reply:
You can always write passes so they operate correctly in the presence of mutation. But we don't want to force it on pass writers. Opaque submodules at final lowering are unlikely to be doing heavy optimization, so it's easier to deal with arbitrary side effects. But even, e.g., finding fusion groups, is currently not correct with mutating things (though mostly this is because of DCE calls). But the problem here is you're adding in this module all the way at the beginning,  before *all* of the optimization passes, and so you're forcing them to, at the very least, know not to remove your module.

Extra outputs is what I would expect to see if it's external.

Reply:
You could also have it not actually be external, but stored on the GraphModule and not actually part of the graph. But passes would need to know to propagate it.

Reply:
hmm yeah, having it actually be external feels better. I'll try that.